### ISSUE TITLE
Demo Architecture Question

### ISSUE THREAD

**Issue by kristiehunter** [hyperledger/aries-cloudagent-python#372]
Hi,

I have a question about the demo architecture that I was hoping I could get some clarification on.

In the documentation there's a clear separation between the Controller API inside the Agent and the Controller Logic outside the agent.  In the recording of walking through this demo, the instructions are that the "Controller" never communicates with the ledger, but one of the first thing the Controller/Wrapper does is call the RegisterDID function, which I see is inherited from the agent.py, but the line before it is the webhook listener (presumably for the Controller piece):

`await agent.listen_webhooks(start_port + 2)`
`await agent.register_did()`

(faber.py)

My team would like to understand this template from the architecture but be able to swap out the Controller Logic piece with another programming language (Node, most likely).  From the demo code, it looks as through the ControllerAPI and the Controller Logic are heavily intertwined in the runner files.

The two-part question is whether or not that's a correct assessment of the code, and if it is if there's a plan to separate it.  Other places in the docs (like the Deployment Model) it says there's no repository for the Controller Logic because it's a basic web server, which is totally fine.  I think we're struggling, though, with what belongs in the API and what belongs in the Logic.  We've been looking for a standalone agent with the API that we can build a Node server to connect to, but haven't found one.

Thanks so much for your help!

**Comment by swcurran** [comment]
Good questions.  The intended design of an ACA-Py based agent is as outlined in the recording - the controller is external to the ACA-Py instance and can be in whatever language you want.  The demo does muddy that up a bit by making the ACA-Py instance a child process of the controller.  In all of the "real" implementations, the controller is external and often written in whatever language makes sense.  The recent work by @amanji has demonstrated this while still using the Alice, Faber example.  See https://github.com/petridishdev/aries-acapy-controllers for examples of 3 controllers all based on a different tech stack (.Net, Angular and node/express). Several examples we have built have often used django, which is still python, but implemented as a separate component running in a separate container.  The Verifiable Credential Open ID Connect implementation that we created (https://github.com/bcgov/vc-authn-oidc) uses .NET as a controller and ACA-Py providing the agent capabilities.

Does that clarify it?
